### PR TITLE
Custom core Request and Response 

### PR DESCRIPTION
**Pull request by MindFlavor**
This PR builds on the previous pipeline work and proposes:

1. Custom `Request` and `Response` structs. While most of them are `http` fields this allows us to be flexible. `Request` accepts a `Body` enum that allows us to support seekable streams in the future. `Response` now does not collect the response stream so, if necessary, it can be passed up to the caller (for example for downloading blobs).
2. A new (temporary) function in `HTTPClient` that operates one the aforementioned structs only. As soon as the other functions are no longer used they will be phased out. Most notable, the *check status* will be moved in a specific policy sometimes in the future (right now it is baked in the transport).
3. A clear per call and per retry policy distinction in the pipeline. 
4. Retry policies (linear, exponential and no retry).
5. The transform policy must be the latest policy. Right now is enforced by the pipeline constructor and asserted at runtime.

## To discuss

* Should the context be mutable?

## Improvement areas

* Policies must either call the following policy or fail. Right now this is not enforced: a policy could skip an arbitrary number of following policies since it gets the whole following pipeline execution to pop from. It could be better if we could somehow enforce this. 

# ⛑️⛑️⛑️⛑️ Important ⛑️⛑️⛑️⛑️ 

This PR temporarily disables the support for Hyper. I have done it to avoid duplicating the work since there are likely to be many suggestions/improvements. It will be reenabled once the PR stabilizes.

### CODE REVIEW

File: sdk/core/src/response.rs
Comment (rylev) [inline sdk/core/src/response.rs:32]: Should this be `with_stream`?

File: sdk/core/src/response.rs
Comment (rylev) [inline sdk/core/src/response.rs:62]: What is this needed for?

File: sdk/cosmos/src/clients/cosmos_client.rs
Comment (rylev) [inline sdk/cosmos/src/clients/cosmos_client.rs:39]: Can you do me a favor and fill in this doc comments? That would be awesome ♥

**Review by rylev — CHANGES_REQUESTED** [review]
Looking great but needs a few changes!

File: sdk/core/src/policies/retry_policies/exponential_retry.rs
Comment (MindFlavor) [reply sdk/core/src/policies/retry_policies/exponential_retry.rs:35]: mmm no, the idea of the code is to update the passed `first_retry_time`. This way the following function call knows that this is not the first retry. I could have returned a tuple `(bool, DateTime<Local>)` but I think it would have been harder to understand. 

I could have used `*first_retry_time = Box::new(Some(Local::now()))` instead of `mem::swap`. Maybe that way is more expressive?

File: sdk/core/src/policies/retry_policies/exponential_retry.rs
Comment (MindFlavor) [reply sdk/core/src/policies/retry_policies/exponential_retry.rs:80]: No, it's not. And you are right, it does not make the code clearer (the intent is self-explanatory enough without having to spell out `continue looping`)!

File: sdk/core/src/request.rs
Comment (MindFlavor) [reply sdk/core/src/request.rs:6]: I think we could get by using an `Arc<Vec<u8>>` (which is for our intent basically the same thing). I like the bytes crate however: it's lean, stable and well known.

File: sdk/core/src/request.rs
Comment (MindFlavor) [reply sdk/core/src/request.rs:48]: BTW do you know if there's a linter that checks this?

File: sdk/core/src/request.rs
Comment (MindFlavor) [reply sdk/core/src/request.rs:50]: Good point. Renamed.

File: sdk/core/src/request.rs
Comment (MindFlavor) [reply sdk/core/src/request.rs:67]: Right now it is used by `prepare_request2` in the cosmos crate. I think once we decouple the authentication from the client and we put it in a policy this won't be necessary anymore.

File: sdk/core/src/response.rs
Comment (MindFlavor) [reply sdk/core/src/response.rs:32]: Yes, maybe even `with_pinned_stream`. 👍

**Comment by MindFlavor** [comment]
I've added support for files and *Seekable streams* (for lack of better term). 
I've only tested the former and it seems to be working as expected.

ASAP I'll try to fix the crates incompatible with WASM.

File: sdk/core/src/http_client.rs
Comment (rylev) [inline sdk/core/src/http_client.rs:122]: doesn't `request.uri()` already return a `url::Url`?

File: sdk/core/src/http_client.rs
Comment (rylev) [inline sdk/core/src/http_client.rs:124]: nit: deconstruct the `(key, value`) here instead of using `.0` notation

File: sdk/core/src/http_client.rs
Comment (rylev) [inline sdk/core/src/http_client.rs:157]: If we need to replace the body for all subsequent retries, shouldn't we create a method like `body_cloned` or something instead of `take_body`?

File: sdk/core/src/http_client.rs
Comment (rylev) [inline sdk/core/src/http_client.rs:211]: Why not delete this?

File: sdk/core/src/policies/retry_policies/exponential_retry.rs
Comment (rylev) [reply sdk/core/src/policies/retry_policies/exponential_retry.rs:35]: I'm referring to `current_retries` not `first_retry_time`

File: sdk/core/src/seekable_stream.rs
Comment (rylev) [inline sdk/core/src/seekable_stream.rs:1]: I'm confused. We worked on removing a dependency on tokio, but now we have a dependency on async_std... I believe all these traits are from the futures crate no?

File: sdk/core/src/seekable_stream.rs
Comment (rylev) [inline sdk/core/src/seekable_stream.rs:35]:
```suggestion
        match self.poll_read(cx, &mut buffer) {
            async_std::task::Poll::Ready(Ok(0)) => {
                    futures::task::Poll::Ready(None)
             } 
             async_std::task::Poll::Ready(Ok(bytes_read)) =>  {
                    let bytes: Bytes = buffer.into();
                    let bytes = bytes.slice(0..bytes_read);
                    futures::task::Poll::Ready(Some(Ok(bytes)))
            }
```

**Review by rylev — CHANGES_REQUESTED** [review]
Let's try to get this merged before adding more features to it. It's looking on it's way, but I had to review a bunch of new code that was added.

File: sdk/core/src/request.rs
Comment (rylev) [reply sdk/core/src/request.rs]: No, I don't unfortunately...

File: sdk/core/Cargo.toml
Comment (ctaggart) [inline sdk/core/Cargo.toml:39]: This is for `ExponentialRetryPolicy`?  So it can have a random delay between retries? Are we sure we want it to always be a core dependency?

File: sdk/core/Cargo.toml
Comment (ctaggart) [inline sdk/core/Cargo.toml:40]: I don't think `async-std` should be a core dependency. It is an alternative to `tokio`.

File: sdk/core/src/request.rs
Comment (ctaggart) [reply sdk/core/src/request.rs:8]: I'd prefer exposing `bytes::Bytes` in our public API.

File: sdk/core/src/seekable_stream.rs
Comment (ctaggart) [inline sdk/core/src/seekable_stream.rs:10]: Conceptually, what is `DynClone` used for here? It is saying that the SeekableStream may be cloned?

File: sdk/core/src/context.rs
Comment (heaths) [inline sdk/core/src/context.rs:11]: Nit: use TODO or BUGBUG to find these easily and fix them before release.

File: sdk/core/src/http_client.rs
Comment (heaths) [inline sdk/core/src/http_client.rs:145]: Can we use a more explanitory name here, like `execute_request_raw` or keep this as `execute_request` and make the other `execute_bytes_request` or `execute_request_with_bytes`? Names like this don't really help developers at a glance know what they do, and we use a similar approach in other languages were overloads aren't supported (JS, Python).

File: sdk/core/src/pipeline.rs
Comment (heaths) [inline sdk/core/src/pipeline.rs:29]: Why make client libraries pass in the retry and transport policies? These should added by default. See https://github.com/Azure/azure-sdk-for-net/blob/8d19a820660c3c34b6ebe7904bd4e7ff895c82f6/sdk/core/Azure.Core/src/Pipeline/HttpPipelineBuilder.cs#L35. This is fairly consistent across languages. If an SDK doesn't want a retry, they can turn it off via ClientOptions.

File: sdk/core/src/pipeline.rs
Comment (heaths) [inline sdk/core/src/pipeline.rs:27]: FWIW, the only thing that should be able to create a pipeline is the client library, e.g. `KeyClient` (Key Vault keys client). However, end users should be able to add their own per-call or per-retry policies. We currently don't allow them to remove built-in ones (ones added by azure_core or the client class).

We probably don't need that here, but should think about how it would work. The previous link should show how we can pass them in. We could either make the ClientOptions classes merge vectors and pass them in, or maybe add 2 parameters here like `user_per_call_policies` and `user_per_retry_policies`. The idea of a builder of these like most (all?) of our languages use might also help make it easier.

File: sdk/core/src/policies/mod.rs
Comment (heaths) [inline sdk/core/src/policies/mod.rs:20]: Policies must receive both the request and response to work. See https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/Pipeline/HttpPipelinePolicy.cs. In .NET, we wrap those in a `Message`. This is critical because some policies act on data from the response itself, like challenge-based auth or the `RetryPolicy` (e.g. `Retry-After` header, which could be either an int (seconds) or string (ISO8601 datetime).

File: sdk/core/src/policies/retry_policies/exponential_retry.rs
Comment (heaths) [inline sdk/core/src/policies/retry_policies/exponential_retry.rs:15]: Re: this and linear retry.

All our SDKs use linear or exponential based on ClientOptions: https://github.com/Azure/azure-sdk-for-net/blob/8d19a820660c3c34b6ebe7904bd4e7ff895c82f6/sdk/core/Azure.Core/src/ClientOptions.cs#L44 and https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/RetryOptions.cs. Thus, how we retry is typically left up to the caller, but the client library can use one or the other, but if we have separate classes we should make them internal and choose the appropriate one based on `RetryOptions`, if having separate classes is easier (see https://github.com/Azure/azure-sdk-for-net/blob/8d19a820660c3c34b6ebe7904bd4e7ff895c82f6/sdk/core/Azure.Core/src/Pipeline/Internal/RetryPolicy.cs#L181-L200 for how we do this in .NET).

File: sdk/core/src/policies/retry_policies/linear_retry.rs
Comment (heaths) [inline sdk/core/src/policies/retry_policies/linear_retry.rs:62]: So I see a policy does get a response, but how would a policy create a response to short-circuit the pipeline and not call `next`? For e.g., to proxy calls.

File: sdk/core/src/policies/transport.rs
Comment (heaths) [inline sdk/core/src/policies/transport.rs:17]: This should be used for a property on ClientOptions for a user (e.g. test infrastructure) to override.

**Review by heaths — APPROVED** [review]
This is a great start and very close to where we need to be, and apart from other comments and ones I left (maybe with a few modifications) could be merged, but maybe open a bug or two as needed for what remains based on comments.

**Comment by heaths** [comment]
For context, here's how we might specify a `MockTransport` in .NET: https://github.com/Azure/azure-sdk-for-net/blob/8d19a820660c3c34b6ebe7904bd4e7ff895c82f6/sdk/keyvault/Azure.Security.KeyVault.Keys/tests/KeyClientTests.cs#L182-L189

In general, our test infrastructure provides some base classes that configure a recording/playback transport:
* Configuring (or variations like this): https://github.com/Azure/azure-sdk-for-net/blob/8d19a820660c3c34b6ebe7904bd4e7ff895c82f6/sdk/keyvault/Azure.Security.KeyVault.Keys/tests/KeysTestBase.cs#L43-L68
* Instrumenting: https://github.com/Azure/azure-sdk-for-net/blob/8d19a820660c3c34b6ebe7904bd4e7ff895c82f6/sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs#L66-L75

The latter can get pretty deep in implementation, but I only wanted to show examples of how the transport policy itself is replaced. A dev, for example, could write their own mock and modify requests and emit responses as they wish.

File: sdk/core/src/http_client.rs
Comment (rylev) [reply sdk/core/src/http_client.rs:145]: There's a comment above explaining this is very temporary. We will rename this to `execute_request` as soon as all requests are moved to using the new pipeline infrastructure. That should hopefully be quickly after merging this PR.

File: sdk/core/src/pipeline.rs
Comment (rylev) [reply sdk/core/src/pipeline.rs:29]: Let's make that change after we merge this PR.

File: sdk/core/src/policies/mod.rs
Comment (rylev) [reply sdk/core/src/policies/mod.rs:20]: The policies do receive both. They receive the request from the caller and then they call the next policy which gives them a response.

File: sdk/core/src/policies/retry_policies/linear_retry.rs
Comment (rylev) [reply sdk/core/src/policies/retry_policies/linear_retry.rs:62]: The policy can just create a response (not sure how easy that is currently but there's nothing preventing us from making it easy) and return that without calling `next[0].send`.

File: sdk/core/src/http_client.rs
Comment (MindFlavor) [reply sdk/core/src/http_client.rs:157]: Yes definitely. Good point! 👍

File: sdk/core/src/http_client.rs
Comment (MindFlavor) [reply sdk/core/src/http_client.rs:211]: It was a workaround introduced by Cameron because reqwest in wasm does not return the version. We do not use it anyway so I've just deleted it.

File: sdk/core/src/seekable_stream.rs
Comment (MindFlavor) [reply sdk/core/src/seekable_stream.rs:1]: Yes I've removed the entire crate from Cargo.toml along with the "file based" stream code.

File: sdk/core/src/seekable_stream.rs
Comment (MindFlavor) [reply sdk/core/src/seekable_stream.rs:35]: Nice! 👍

File: sdk/core/Cargo.toml
Comment (MindFlavor) [reply sdk/core/Cargo.toml:39]: Yes exactly I modeled the retry based on this: [https://docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#retry-mechanism-3](https://docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#retry-mechanism-3). BTW rand 0.8 broke wasm so I downgraded to 0.7. Not sure what's different though.

File: sdk/core/Cargo.toml
Comment (MindFlavor) [reply sdk/core/Cargo.toml:40]: Yes you are right. As discussed yesterday I've scrapped the code that required `aync-std` (file based input stream).

File: sdk/core/src/seekable_stream.rs
Comment (MindFlavor) [reply sdk/core/src/seekable_stream.rs:10]: Yes, the stream must be clonable because reqwest (and I suppose any other client http implementation) takes ownership of it, but we need ownership back for subsequent retries. 

I've added a single function (`reset`) that replaces the `AsyncSeek` trait because we do not need arbitrary seeking, just to able to start from the beginning.

This is an abomination, I agree (also I do not know what could possibly implement it). In the previous iteration of the code I've implemented the trick of saving the original stream so we could replay the saved copy during retries. In other words, the original stream would only be executed once: our code would send the "saved" copy if a retry is needed. The trait than didn't need to be `clone`.

File: sdk/core/src/context.rs
Comment (MindFlavor) [reply sdk/core/src/context.rs:11]: Good idea. 👍

File: sdk/core/src/policies/retry_policies/linear_retry.rs
Comment (MindFlavor) [reply sdk/core/src/policies/retry_policies/linear_retry.rs:62]: To make this easier I've added a simple struct that implements `Stream` based on a byte buffer (called `BytesStream`). A policy that wants to short-circuit the pipeline can then construct a response easily (I've used it in the mock wasm implementation for reqwest).

**Comment by heaths** [comment]
Was this ready to merge? I have the telemetry policy ready to go - mostly, I'd need to rebase on this and tweak it a little. I can do that now and rebase later, but if this is ready to merge it'd be nice to get it in.

Per discussion with @rylev I'll use composition for diagnostic and client options. We should probably standardize on that across the board, so for a KeyClientOptions for a KeyClient client class, it'd likely look something like this:

```rust
pub enum ServiceLevel {
  V7_2,
  V7_1,
  V7_0,
}

pub struct KeyClientOptions {
  pub serviceLevel: ServiceLevel,
  pub clientOptions: ClientOptions,
}
```

**Comment by heaths** [comment]
Ultimately we'll need to implement a trait to stringify the service level (e.g. turn it into "7.2", for example). Would it be best to implement `Into`?

**Comment by MindFlavor** [comment]
> Was this ready to merge? I have the telemetry policy ready to go - mostly, I'd need to rebase on this and tweak it a little. I can do that now and rebase later, but if this is ready to merge it'd be nice to get it in.

We are waiting for Ryan's approval (he's been busy the past couple of weeks). There are a few tweaks suggested by Jeffrey but we'll address them in a separate PR.

File: sdk/core/src/http_client.rs
Comment (rylev) [inline sdk/core/src/http_client.rs:10]: Why are we allowing this?

File: sdk/core/src/http_client.rs
Comment (rylev) [reply sdk/core/src/http_client.rs:126]: Ping!

**Review by rylev — APPROVED** [review]
There's still some comments that haven't yet been addressed, but I'll merge this just so we can get it in.

**Pull request merged**
